doc/man/gnunet-resolver.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd January 4, 2012
.Dt GNUNET-RESOLVER 1
.Os
.Sh NAME
.Nm gnunet-resolver
.Nd built-in GNUnet stub resolver
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl l Ar LOGFILE | Fl -logfile= Ns Ar LOGFILE
.Op Fl r | -reverse
.Op Fl v | -version
.Ao Ar hostname Ac
.Sh DESCRIPTION
.Nm
allows you to use the built-in GNUnet stub resolver.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl l Ar LOGFILE | Fl -logfile= Ns Ar LOGFILE
Configure logging to write logs to LOGFILE.
.It Fl r | -reverse
Perform a reverse lookup.
.It Fl v | -version
Print GNUnet version number.
.El
.\".Sh EXAMPLES
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Sat, 3 Apr 2021 09:40:46 +0200 (08fe7735b)
handbook: Add section on commit messages and developer branches - Martin Schanzenbach

Thu, 1 Apr 2021 00:41:38 +0200 (9f340d5cc)
SETU: added some more perf measurement point for var message size - Elias Summermatter

Thu, 1 Apr 2021 00:15:43 +0200 (12bb61317)
SETU: fixed some bug in rand element gen and added variable payload for perfmance - Elias Summermatter

Wed, 31 Mar 2021 11:49:10 +0200 (902331756)
SETU: Added new operation parameters for performance tests - Elias Summermatter

Wed, 31 Mar 2021 11:16:38 +0200 (b5660e0a3)
SETU: improved tests - Elias Summermatter

Tue, 30 Mar 2021 19:26:40 +0200 (ca4c9eae2)
UTIL/SCALARPRODUCT: Do not build ECC dlog or scalarproduct with incompatible libgcrypt versions (>=1.9.0) - Martin Schanzenbach

Mon, 22 Mar 2021 20:06:33 +0100 (a42d7df64)
fix #6808: gnunet-bcd card generation - Martin Schanzenbach

Thu, 18 Mar 2021 18:44:46 +0100 (7d85c27e9)
Fixed setu tests and state maschine - Elias Summermatter

Thu, 18 Mar 2021 17:55:24 +0100 (8cd1e1929)
Fixed some bug in the statemaschin implementation of setu - Elias Summermatter

Thu, 18 Mar 2021 16:07:40 +0100 (c4ccd7550)
Setu tests use gnunet internal functions to generate randum elements - Elias Summermatter

Thu, 18 Mar 2021 11:48:42 +0100 (f9ae0c23c)
Added working test with randum set - Elias Summermatter

Thu, 18 Mar 2021 08:45:31 +0100 (0d6237ae0)
Added some randum element generation test for setu - Elias Summermatter

Sun, 14 Mar 2021 14:04:00 +0100 (b47586e76)
add packaging notice; fix #5633 - Martin Schanzenbach

Tue, 9 Mar 2021 18:26:28 +0100 (0b2c0f22f)
IDENTITY: Fix wrong key construction for anonymous ECDSA identity - Martin Schanzenbach

Tue, 2 Mar 2021 17:16:18 +0100 (45ceb4fdf)
add base32 encoder/decoder - Christian Grothoff

Sun, 28 Feb 2021 13:51:25 +0100 (4cffc642c)
build: do not redefine AM_MAKEINFOHTMLFLAGS - Martin Schanzenbach

Sun, 28 Feb 2021 11:15:55 +0100 (eb8ca760c)
GANA: Use GANA generated header for GNS record types. Update handbook - Martin Schanzenbach

Tue, 23 Feb 2021 14:29:47 +0100 (d528f6cfb)
HELLO: Fix netdb.h nonsense with h_addr - Martin Schanzenbach

Tue, 23 Feb 2021 10:42:57 +0100 (74a234f0b)
HANDBOOK: Add some info on TNG - Martin Schanzenbach

Sun, 21 Feb 2021 20:56:47 +0100 (05040a491)
HELLO: Add test for NG API; fix extraction more - Martin Schanzenbach

Sun, 21 Feb 2021 16:44:51 +0100 (783a12b3c)
HELLO-TNG: Fix parsing; sill broken - Martin Schanzenbach

Sun, 21 Feb 2021 11:46:16 +0100 (f5439c229)
TNG: Add queue update handling - Martin Schanzenbach

Sun, 21 Feb 2021 10:53:34 +0100 (a3971a93c)
Add struct names for FS events. #6743 - Martin Schanzenbach

Thu, 18 Feb 2021 16:51:47 +0100 (8095b1a1b)
SetU move check to union - Elias Summermatter

Wed, 17 Feb 2021 18:00:43 +0100 (08adc88b4)
Setu integrated new phase full receiving as described in rfc draft - Elias Summermatter

Wed, 17 Feb 2021 17:32:23 +0100 (d5bf65f2b)
Renamed Phase Expect IBF Cont -> Expect IBF Last - Elias Summermatter

Wed, 17 Feb 2021 17:26:35 +0100 (d852cb5a6)
Renamed Phase Done -> Finished - Elias Summermatter

Wed, 17 Feb 2021 17:04:15 +0100 (be8897e14)
Renamed Phase inventory passive -> passive decoding - Elias Summermatter

Wed, 17 Feb 2021 15:57:46 +0100 (333459251)
Renamed Phase inventory active -> active decoding & Added idea folder to git ignore - Elias Summermatter

Sat, 30 Jan 2021 16:36:14 -0500 (5b152adc0)
use AS_VERSION_COMPARE instead of AX_COMPARE_VERSION - Thien-Thi Nguyen

Sat, 30 Jan 2021 08:37:12 -0500 (de1b87ea3)
fix #6722: use AX_COMPARE_VERSION - Thien-Thi Nguyen

Sat, 30 Jan 2021 12:41:15 +0100 (a5f200d53)
fix #6690 - Christian Grothoff

Thu, 28 Jan 2021 22:28:01 +0100 (a2169368b)
implement #6716 - Christian Grothoff

Thu, 7 Jan 2021 22:22:48 +0100 (4769344a7)
convert to GMT, not localtime in GNUNET_TIME_year_to_time - Christian Grothoff

Thu, 7 Jan 2021 10:26:54 +0100 (5b10ad755)
properly document return value of GNUNET_PROGRAM_run - Christian Grothoff

Sat, 2 Jan 2021 13:47:16 +0100 (eab08cec9)
do not output synthetic paths - Christian Grothoff

Sat, 2 Jan 2021 11:24:26 +0100 (0d4337da2)
fix /proc/PID/maps parser format string - Christian Grothoff

Thu, 31 Dec 2020 09:12:43 +0100 (ebd853c83)
do not use atomic write for configuration file, we may not have permissions to do this, also should not be necessary - Christian Grothoff

Wed, 30 Dec 2020 20:58:55 +0900 (c0465ff92)
RECLAIM: Fix quirky OIDC address handling - Martin Schanzenbach

Sat, 26 Dec 2020 14:15:58 +0900 (75a3a90fb)
RECLAIM: automatically purge dangling references - Martin Schanzenbach

Sat, 19 Dec 2020 18:43:38 +0100 (3636ea628)
change GNUNET_DISK_fn_write() to always do atomic writes and to NOT overwrite existing files; also change the return value to not return the size of the written file but GNUNET_OK on success, and integrate creating the directory if needed; breaks API, hence bumping libgnunetutil version - Christian Grothoff

Tue, 15 Dec 2020 04:54:32 -0500 (407765df1)
include Configuration Handbook in gnunet.info - Thien-Thi Nguyen

Tue, 15 Dec 2020 03:43:27 -0500 (db9d81947)
add pxref to ‘Config file format’ - Thien-Thi Nguyen

Tue, 15 Dec 2020 03:26:26 -0500 (52986bf33)
[doc] Add section "Config file format" - Thien-Thi Nguyen

Wed, 9 Dec 2020 23:00:07 +0100 (7d9592f20)
add GNUNET_CRYPTO_hash_context_copy() function - Christian Grothoff

Tue, 8 Dec 2020 16:50:27 +0900 (7ce054864)
RECLAIM: Return userinfo claims from cache - Martin Schanzenbach

Fri, 4 Dec 2020 17:36:35 +0100 (1346641a1)
gnunet-crypto-tvg: fix bogus label - Florian Dold

Fri, 4 Dec 2020 17:22:38 +0100 (4c10ff3b4)
gnunet-crypto-tvg: add verification mode - Florian Dold

Fri, 4 Dec 2020 14:52:45 +0100 (1a3070b15)
gnunet-crypto-tvg: output JSON, fix memleaks - Florian Dold

Sat, 21 Nov 2020 22:58:44 +0000 (d2e10ef67)
rps: fix another 'format not a string literal' warning - Daniel Golle

Mon, 16 Nov 2020 23:57:08 +0100 (91e8c2ff2)
rps: deal with currently unused code - Julius Bünger


src/util/gnunet-resolver.c
/*
     This file is part of GNUnet.
     Copyright (C) 2010 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file util/gnunet-resolver.c
 * @brief tool to test resolver
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_resolver_service.h"

#define GET_TIMEOUT GNUNET_TIME_relative_multiply (GNUNET_TIME_UNIT_SECONDS, 5)

/**
 * Flag for reverse lookup.
 */
static int reverse;


/**
 * Prints each hostname obtained from DNS.
 *
 * @param cls closure (unused)
 * @param hostname one of the names for the host, NULL
 *        on the last call to the callback
 */
static void
print_hostname (void *cls,
                const char *hostname)
{
  (void) cls;
  if (NULL == hostname)
    return;
  fprintf (stdout,
           "%s\n",
           hostname);
}


/**
 * Callback function to display address.
 *
 * @param cls closure (unused)
 * @param addr one of the addresses of the host, NULL for the last address
 * @param addrlen length of the address
 */
static void
print_sockaddr (void *cls,
                const struct sockaddr *addr,
                socklen_t addrlen)
{
  (void) cls;
  if (NULL == addr)
    return;
  fprintf (stdout,
           "%s\n",
           GNUNET_a2s (addr,
                       addrlen));
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  const struct sockaddr *sa;
  socklen_t salen;
  struct sockaddr_in v4;
  struct sockaddr_in6 v6;

  (void) cls;
  (void) cfgfile;
  (void) cfg;
  if (NULL == args[0])
    return;
  if (! reverse)
  {
    GNUNET_RESOLVER_ip_get (args[0],
                            AF_UNSPEC,
                            GET_TIMEOUT,
                            &print_sockaddr,
                            NULL);
    return;
  }

  sa = NULL;
  memset (&v4, 0, sizeof(v4));
  v4.sin_family = AF_INET;
#if HAVE_SOCKADDR_IN_SIN_LEN
  v4.sin_len = sizeof(v4);
#endif
  if (1 == inet_pton (AF_INET,
                      args[0],
                      &v4.sin_addr))
  {
    sa = (struct sockaddr *) &v4;
    salen = sizeof(v4);
  }
  memset (&v6, 0, sizeof(v6));
  v6.sin6_family = AF_INET6;
#if HAVE_SOCKADDR_IN_SIN_LEN
  v6.sin6_len = sizeof(v6);
#endif
  if (1 == inet_pton (AF_INET6,
                      args[0],
                      &v6.sin6_addr))
  {
    sa = (struct sockaddr *) &v6;
    salen = sizeof(v6);
  }
  if (NULL == sa)
  {
    fprintf (stderr,
             "`%s' is not a valid IP: %s\n",
             args[0],
             strerror (errno));
    return;
  }
  GNUNET_RESOLVER_hostname_get (sa, salen,
                                GNUNET_YES,
                                GET_TIMEOUT,
                                &print_hostname,
                                NULL);
}


/**
 * The main function to access GNUnet's DNS resolver.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_flag ('r',
                               "reverse",
                               gettext_noop ("perform a reverse lookup"),
                               &reverse),
    GNUNET_GETOPT_OPTION_END
  };
  int ret;

  if (GNUNET_OK !=
      GNUNET_STRINGS_get_utf8_args (argc, argv,
                                    &argc, &argv))
    return 2;

  ret = (GNUNET_OK ==
         GNUNET_PROGRAM_run (argc, argv,
                             "gnunet-resolver [hostname]",
                             gettext_noop ("Use built-in GNUnet stub resolver"),
                             options,
                             &run, NULL)) ? 0 : 1;
  GNUNET_free_nz ((void *) argv);
  return ret;
}


/* end of gnunet-resolver.c */
